libnopegl/src/text_external.c

```diff
@@ -0,0 +1,749 @@
+/*
+ * Copyright 2023 Nope Project
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "config.h"
+
+#if HAVE_TEXTLIBS
+#include <hb.h>
+#include <hb-ft.h>
+#include <ft2build.h>
+#include FT_FREETYPE_H
+#include <fribidi.h>
+#endif
+
+#include "atlas.h"
+#include "darray.h"
+#include "hmap.h"
+#include "log.h"
+#include "memory.h"
+#include "nopegl.h"
+#include "text.h"
+#include "utils.h"
+
+#if HAVE_TEXTLIBS
+struct text_external {
+    FT_Library ft_library;
+    struct darray ft_faces; // FT_Face (hidden pointer)
+    struct darray hb_fonts; // hb_font_t*
+    struct atlas *atlas;
+};
+
+static int load_font(struct text *text, const char *font_file)
+{
+    int ret = 0;
+    struct text_external *s = text->priv_data;
+
+    /* This limitation simplifies the UID computation in GLYPH_UID_STRING() */
+    if (ngli_darray_count(&s->ft_faces) == 0xff) {
+        LOG(ERROR, "maximum number of fonts reached (256)");
+        return NGL_ERROR_LIMIT_EXCEEDED;
+    }
+
+    FT_Face ft_face = NULL;
+    hb_font_t *hb_font = NULL;
+
+    FT_Error ft_error = FT_New_Face(s->ft_library, font_file, 0, &ft_face);
+    if (ft_error) {
+        LOG(ERROR, "unable to initialize FreeType with font %s", font_file);
+        ret = NGL_ERROR_EXTERNAL;
+        goto end;
+    }
+
+    if (!ngli_darray_push(&s->ft_faces, &ft_face)) {
+        ret = NGL_ERROR_MEMORY;
+        goto end;
+    }
+
+    const int32_t pt_size = 54;
+    const FT_F26Dot6 chr_w = NGLI_I32_TO_I26D6(pt_size);
+    const FT_F26Dot6 chr_h = NGLI_I32_TO_I26D6(pt_size);
+    const FT_UInt h_res = 96;
+    const FT_UInt v_res = 96;
```

Comment:
Can you add a comment about which size this corresponds to ?

Reply:
Changed to:
```c
    const FT_F26Dot6 chr_w = NGLI_I32_TO_I26D6(pt_size); // nominal width in 26.6
    const FT_F26Dot6 chr_h = NGLI_I32_TO_I26D6(pt_size); // nominal height in 26.6
    const FT_UInt h_res = 96; // horizontal resolution in dpi
    const FT_UInt v_res = 96; // vertical resolution in dpi
```

Reply:
Thanks